packages/dev/core/src/Engines/WebGL/webGLRenderTargetWrapper.ts

```diff
@@ -75,23 +75,46 @@ export class WebGLRenderTargetWrapper extends RenderTargetWrapper {
      * Binds a texture to this render target on a specific attachment
      * @param texture The texture to bind to the framebuffer
      * @param attachmentIndex Index of the attachment
-     * @param faceIndex The face of the texture to render to in case of cube texture
+     * @param faceIndexOrLayer The face or layer of the texture to render to in case of cube texture or array texture
      * @param lodLevel defines the lod level to bind to the frame buffer
      */
-    private _bindTextureRenderTarget(texture: InternalTexture, attachmentIndex: number = 0, faceIndex: number = -1, lodLevel: number = 0) {
+    private _bindTextureRenderTarget(texture: InternalTexture, attachmentIndex: number = 0, faceIndexOrLayer?: number, lodLevel: number = 0) {
         if (!texture._hardwareTexture) {
             return;
         }
 
-        const gl = this._context;
         const framebuffer = this._framebuffer;
 
         const currentFB = this._engine._currentFramebuffer;
         this._engine._bindUnboundFramebuffer(framebuffer);
-        const attachment = (<any>gl)[this._engine.webGLVersion > 1 ? "COLOR_ATTACHMENT" + attachmentIndex : "COLOR_ATTACHMENT" + attachmentIndex + "_WEBGL"];
-        const target = faceIndex !== -1 ? gl.TEXTURE_CUBE_MAP_POSITIVE_X + faceIndex : gl.TEXTURE_2D;
 
-        gl.framebufferTexture2D(gl.FRAMEBUFFER, attachment, target, texture._hardwareTexture.underlyingResource, lodLevel);
+        if (this._engine.webGLVersion > 1) {
+            const gl = this._context as WebGL2RenderingContext;
+
+            const attachment = (<any>gl)["COLOR_ATTACHMENT" + attachmentIndex];
+            if (texture.is2DArray || texture.is3D) {
+                faceIndexOrLayer = faceIndexOrLayer ?? this.layerIndices?.[attachmentIndex] ?? 0;
+                gl.framebufferTextureLayer(gl.FRAMEBUFFER, attachment, texture._hardwareTexture.underlyingResource, lodLevel, faceIndexOrLayer);
+            } else if (texture.isCube) {
+                // if face index is not specified, try to query it from faceIndices
+                // default is face 0
+                faceIndexOrLayer = faceIndexOrLayer ?? this.faceIndices?.[attachmentIndex] ?? 0;
+                gl.framebufferTexture2D(gl.FRAMEBUFFER, attachment, gl.TEXTURE_CUBE_MAP_POSITIVE_X + faceIndexOrLayer, texture._hardwareTexture.underlyingResource, lodLevel);
+            } else {
+                gl.framebufferTexture2D(gl.FRAMEBUFFER, attachment, gl.TEXTURE_2D, texture._hardwareTexture.underlyingResource, lodLevel);
+            }
+        } else {
+            // Default behavior (WebGL)
+            const gl = this._context;
+
+            faceIndexOrLayer = faceIndexOrLayer ?? -1;
+
+            const attachment = (<any>gl)["COLOR_ATTACHMENT" + attachmentIndex + "_WEBGL"];
+            const target = faceIndexOrLayer !== -1 ? gl.TEXTURE_CUBE_MAP_POSITIVE_X + faceIndexOrLayer : gl.TEXTURE_2D;

```

Comment:
`setsetLayerAndFaceIndex` defaults to -1, so I made changes there too; I think I left it like this because it was backwards compatible. However, I don't think `_bindTextureRenderTarget` was called previously with the `faceIndex` parameter set, and nothing else relies on `layers` being set to -1, so I changed everything to optional parameters.